node/src/reactor/participating.rs

```diff
@@ -539,55 +540,181 @@ impl reactor::Reactor for Reactor {
+                    trace!(%post_state_hash, ?execution_effect);
+                    initial_pre_state = ExecutionPreState::new(
+                        post_state_hash.into(),
+                        0,
+                        Default::default(),
+                        Default::default(),
+                    );
+                    finalized_block = FinalizedBlock::new(
+                        BlockPayload::default(),
+                        Some(EraReport::default()),
+                        genesis_timestamp,
+                        EraId::from(0u64),
+                        0,
+                        PublicKey::System,
+                    );
+                }
+                ActivationPoint::EraId(upgrade_era_id) => {
+                    let upgrade_block_header = storage
+                        .read_switch_block_header_by_era_id(upgrade_era_id.saturating_sub(1))?
+                        .ok_or(Error::NoSuchSwitchBlockHeaderForUpgradeEra { upgrade_era_id })?;
+                    // If it's not an emergency upgrade and there is a block higher than ours, bail
+                    // because we will overwrite an existing blockchain.
+                    if chainspec.protocol_config.last_emergency_restart
+                        != Some(upgrade_block_header.era_id())
+                    {
+                        if let Some(preexisting_block_header) = storage
+                            .read_block_header_by_height(upgrade_block_header.height() + 1)?
+                        {
+                            return Err(Error::NonEmergencyUpgradeWillClobberExistingBlockChain {
+                                preexisting_block_header: Box::new(preexisting_block_header),
+                            });
+                        }
+                    }
+                    let chainspec = chainspec_loader.chainspec();
+                    let global_state_update = chainspec.protocol_config.get_update_mapping()?;
+                    let UpgradeSuccess {
+                        post_state_hash,
+                        execution_effect,
+                    } = contract_runtime.commit_upgrade(UpgradeConfig::new(
+                        (*upgrade_block_header.state_root_hash()).into(),
+                        upgrade_block_header.protocol_version(),
+                        chainspec.protocol_version(),
+                        Some(chainspec.wasm_config),
+                        Some(chainspec.system_costs_config),
+                        Some(chainspec.protocol_config.activation_point.era_id()),
+                        Some(chainspec.core_config.validator_slots),
+                        Some(chainspec.core_config.auction_delay),
+                        Some(chainspec.core_config.locked_funds_period.millis()),
+                        Some(chainspec.core_config.round_seigniorage_rate),
+                        Some(chainspec.core_config.unbonding_delay),
```

Comment:
Chatting with @Fraser999 , consider having `CoreConfig` as an argument.

Reply:
There's a lot of good engineering practices against doing that:
* you should be passing only what's needed - that way it makes it easier to test (one doesn't have to mock all other fields); you make it easier to extend (you're not hardcoding the structure to use the config struct), it's more flexible (if we refactor `CoreConfig` we would have to also refactor `CommitUpgrade`.
* you won't be copying as much data (rather than a couple of fields, it would need to copy whole structure).
* `CoreConfig` is a type from "another level" of the stack, different context you might say - it's a configuration layer rather than EE specific thing.

Reply:
The code was adapted from some of Fraser's ChainSpec code [here][1].  We identified it as having too many arguments and proposed it take a data structure instead.

[1]: https://github.com/xcthulhu/casper-node/blob/1c1d2564d40d0e72472ca16cab9bb35fe180245d/node/src/components/chainspec_loader.rs#L534-L547